### ISSUE TITLE
Support for Scala.js

### ISSUE THREAD

**Issue by Sciss** [scala-stm/scala-stm#12]
I'm trying to use Scala-STM in a Scala.js context. At the glance of it, all compiles, but there are deferred linker errors, I'm not sure why they don't show up in the first place.

Here is a work in progress: https://github.com/scala-stm/scala-stm/tree/sjs

@nbronson can you have a look, it seems `AtomicLongFieldUpdater` is not supported; looking into its source, it seems to do some reflection based stuff. Is there any reason that we need reflection to access the `meta` field? The uses are in `CCSTMRefs`. For example:

```scala
  private class FloatRef(@volatile var data: Float) extends BaseRef[Float] {
    @volatile var meta = 0L
    def metaCAS(m0: Long, m1: Long): Boolean = floatUpdater.compareAndSet(this, m0, m1)
    def newMetaUpdater: AtomicLongFieldUpdater[FloatRef] =
      AtomicLongFieldUpdater.newUpdater(classOf[FloatRef], "meta")
  }
```
This is used in `private val floatUpdater`, and we only ever call `floatUpdater.compareAndSet(this, m0, m1)`. Seems like we could avoid using reflection at all ... ?

**Comment by Sciss** [comment]
I can see that it tests VM capabilities, etc. so I guess it's there for a good (performance) reason. Could you @nbronson imagine that there is another version of `CCSTMRefs` for JS that uses more generic API that doesn't rely on the JVM? As far as I can see `AtomicLong` is supported, so could we just use an instance of `AtomicLong` instead of `@volatile var meta: Long`?

**Comment by Sciss** [comment]
Here are the issues so far:

- cannot use `OptManifest`, same thing as with Dotty; so we could try to go bleeding edge and only support SJS for Scala 3. Kind of wacky, I think. Or make yet another version of RefFactory without using manifests
- methods like `Float.floatToRawIntBits` are not supported. I think they should be used still on the JVM, because they are the correct methods to use. On SJS we have to fall back to `Float.floatToIntBits`
- `Thread#yield` and `Object#wait` are not supported, as there is no real multi-threadedness on JS. That means also copying a lot of classes and removing those calls (in the expectation and hope that they will never kick in because there simply aren't competing threads).

**Comment by Sciss** [comment]
For testing purposes, the branch now builds and can be `publishLocal`'ed

**Comment by Sciss** [comment]
I will publish a new version 0.11.0 that also supports Scala.js on 2.13. Since JVM is basically unchanged, I don't think it needs review; it is understood that SJS support is absolutely experimental at this stage.

**Comment by Sciss** [comment]
fixed in v0.11.0
